src/instrument.ts

```diff
@@ -0,0 +1,249 @@
+/*!
+ * Copyright 2024 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  SEMATTRS_DB_NAME,
+  SEMATTRS_DB_STATEMENT,
+  SEMATTRS_DB_SYSTEM,
+  SEMATTRS_DB_SQL_TABLE,
+} from '@opentelemetry/semantic-conventions';
+
+import {
+  Span,
+  SpanStatusCode,
+  trace,
+  INVALID_SPAN_CONTEXT,
+  SpanAttributes,
+  TimeInput,
+  TracerProvider,
+  Link,
+  Exception,
+  SpanContext,
+  SpanStatus,
+  SpanKind,
+} from '@opentelemetry/api';
+
+const optedInPII: boolean =
+  process.env.SPANNER_ENABLE_EXTENDED_TRACING === 'true';
+
+interface SQLStatement {
+  sql: string;
+}
+
+interface observabilityOptions {
+  tracerProvider: TracerProvider;
+  enableExtendedTracing: boolean;
+}
+
+export type {observabilityOptions as ObservabilityOptions};
+
+const TRACER_NAME = 'cloud.google.com/nodejs/spanner';
+const TRACER_VERSION = '7.14.0'; // Manually hard coded, TODO: remove
+
+/**
+ * getTracer fetches the tracer from the provided tracerProvider.
+ * @param {TracerProvider} [tracerProvider] -- optional custom tracer provider
+ * to use for fetching the tracer. If not provided, the global provider will be used.
+ *
+ * @returns {Tracer} The tracer instance associated with the provided or global provider.
+ */
+export function getTracer(tracerProvider?: TracerProvider) {
+  if (tracerProvider) {
+    return tracerProvider.getTracer(TRACER_NAME, TRACER_VERSION);
+  }
+  // Otherwise use the global tracer.
+  return trace.getTracer(TRACER_NAME, TRACER_VERSION);
+}
+
+interface traceConfig {
+  sql?: string | SQLStatement;
+  tableName?: string;
+  dbName?: string;
+  opts?: observabilityOptions;
+  that?: Object;
+}
+
+const SPAN_NAMESPACE_PREFIX = 'cloud.google.com/nodejs/spanner';
```

Comment:
Why are we adding this prefix in all the traces ?

Reply:
To correctly identify all the spans as originating from this library and namespaces correctly. For example all gRPC spans emanating from the gRPC instrumentation are prefixed with "grpc.google.spanner.v1" and for libraries with proper instrumentation like Go, every span is prefixed with the proper "cloud.google.com/go/spanner".

Reply:
That can be done via `Scope name` as we are adding this in instrumentation scope. 

@harshachinta `for libraries with proper instrumentation like Go, every span is prefixed with the proper "cloud.google.com/go/spanner".` Is it really happening ?

Reply:
Yes, and historically it could be because opencensus didn't have the scope name support and it was hard to identify which library emits a particular trace.
Whereas in opentelemetry we have a scope name that can be set. So lets **not add the full name as prefix** to every span in python and node since it could make it hard to read the spans.

Reply:
@harshachinta @surbhigarg92 the reason for using the namespace is more to aid in searching for spans than the use of scope name: most tracing tools like Google Cloud Trace, Zipkin, Jaeger have a search by span name prefix. It'll be impossible to find spaces produced by a library without their namespace. The use now of "scope" will be a drastic change for tools and visualizers. I highly recommend we stick with historical precedent because even the gRPC OpenTelemetry instrumentation that's modernized uses full namespace.

Reply:
I suppose we can reduce it to "CloudSpanner." (like java-spanner and python-spanner do) but that'll also mean that we couldn't easily find spans from a language or library and systems like Google Cloud Trace do not support searching by attributes but rather by search by name
<img width="1462" alt="Screenshot 2024-08-28 at 3 02 35 PM" src="https://github.com/user-attachments/assets/17b0c404-d35a-48eb-96d1-233c55181c72">

Reply:
@surbhigarg92 @harshachinta  I think that to avoid ruffling too many feathers, we can use `CloudSpanner.` as the prefix to match java-spanner and python-spanner, but we are definitely going to have to make a standardization decision and discussion later on across all Cloud Spanner libraries, or have to change go-spanner which uses the fully qualified names.